tests/conftest.py

```diff
@@ -36,29 +42,156 @@
+            return
+
+        self.EMULATOR = self._get_emulator()
+        self.EMULATOR.start()
+
+    def stop(self) -> None:
+        """Stop the emulator and delete its profile dir.
+
+        Not doing anything for master process as the emulator was not started.
+        """
+        if self._is_master():
+            return
+
+        self.EMULATOR.stop()
+        self._delete_profile()
+
+    def get_port(self) -> int:
+        """Get a unique port for this worker process on which it can run.
+
+        Guarantees to be unique because each worker has a different name.
+        gw0=>20000, gw1=>20111, gw2=>20222, etc.
+
+        Assumes there will not be more than 10 workers.
+        """
+        worker_id = self._get_worker_id()
+        assert worker_id is not None  # shouldn't be run in master process
+        port = "20" + worker_id[-1] * 3
+        return int(port)
+
+    def _get_emulator(self) -> Emulator:
+        """Get the emulator instance for current model."""
+        if self.model == "core":
+            return CoreEmulator(
+                self.EMU_CORE,
+                profile_dir=self._get_profile(),
+                port=self.get_port(),
+                headless=self.headless,
+            )
+        else:
+            # TODO: does not support port yet
+            return LegacyEmulator(
+                self.EMU_CORE,
+                profile_dir=str(self._get_profile()),
+                headless=self.headless,
+            )
+
+    def _delete_profile(self) -> None:
+        """Clean the emulator profile dir."""
+        shutil.rmtree(self._get_profile())
+
+    def _get_profile(self) -> Path:
+        """Profile is a unique dir according to the port number."""
+        return Path("/var/tmp") / str(self.get_port())
+
+    def _get_worker_id(self) -> Union[str, None]:
+        """Workers are named `gw0`, `gw1` etc.
+
+        Returns None for the master/controller process.
+        """
+        return os.getenv("PYTEST_XDIST_WORKER")
 
-@pytest.fixture(scope="session")
+    def _is_master(self) -> bool:
+        """Is this the master/controller process?
+
+        We do not want to start/stop emulators in that case.
+        """
+        return self._get_worker_id() is None
+
+
+@pytest.fixture(scope="function")
 def _raw_client(request: pytest.FixtureRequest) -> Client:
-    path = os.environ.get("TREZOR_PATH")
     interact = int(os.environ.get("INTERACT", 0))
-    if path:
-        try:
-            transport = get_transport(path)
-            return Client(transport, auto_interact=not interact)
-        except Exception as e:
-            request.session.shouldstop = "Failed to communicate with Trezor"
-            raise RuntimeError(f"Failed to open debuglink for {path}") from e
-
+    # In case tests run in parallel, each process has its own path to the emulator.
+    if EMU is not None:
+        path = f"udp:127.0.0.1:{EMU.get_port()}"
+        return _client_from_path(request, path, interact)
```

Comment:
`Emulator` will set up `emu.client` for you which you can return directly -- but perhaps you'll need to add explicit `auto_interact` argument to `Emulator` so that the `debug` value is decoupled from it. We want to be able to run emulator in debug mode (to get console output) but keep manual interaction.

Reply:
Done in [00aa807](https://github.com/trezor/trezor-firmware/pull/2462/commits/00aa8071e1d9f549436b90ea80bbf6d813bfb691)